CCI refactor

sebageek opened the pull request:
Will incorporate all avocado/CCI changes (including bm/direct port binding/trunk support) and a lot of driver cleanup.

sebageek committed:
Move AllocationsManager to db/models.py

When we start to add other models then we can do this in a somewhat
standard place and move the old ones there as well (for as long as we
still need them).

sebageek committed:
Refactor ACI config parsing and usage

The driver now uses oslo.config to parse hostgroups, fixed bindings and
address scopes. Also we now use global singletons in places where they
already exist instead of passing cfg.CONF off to classes and then use
self.conf.

sebageek committed:
Disable config reload via SIGHUP

This feature is not used by us, also we're going to go with immutable
config maps for deploying the driver.

sebageek committed:
Move switch_from_local_link to common.py, rename it

Other places need this function as well, no need to keep this utility
function in the CiscoACIMechanismDriver.

sebageek committed:
Add config opts to split up contracts datastructure

The datastructure for address scope provided/consumed contracts is
needlessly complicated. Moving forward we can use consumed_contracts and
provided_contracts instead. For the time being contracts will be parsed
and extend those two lists.

sebageek committed:
Move network sync from RPC loop into separate function

This makes the network sync easily callable via the backdoor and will
also make it easier to call it via API.

joker-at-work commented on networking_aci/plugins/ml2/drivers/mech_aci/agent/aci_agent.py:
```diff
@@ -147,13 +147,6 @@ def _check_and_handle_signal(self):
             LOG.info(_LI("Agent caught SIGTERM, quitting daemon loop."))
             self.run_daemon_loop = False
             self.catch_sigterm = False
-        if self.catch_sighup:
```
Would it make sense to also remove the `self.catch_sighup` attribute?

sebageek replied:
Yes

joker-at-work commented on networking_aci/plugins/ml2/drivers/mech_aci/common.py:
```diff
@@ -72,3 +74,26 @@ def get_address_scope_name(self, context, subnet_pool_id):
+            else:
+                LOG.error("Local information %s is invalid in binding profile %s.",
+                          lli, binding_profile)
+        except ValueError:
+            LOG.info("binding Profile %s cannot be parsed", binding_profile)
```
Would it make sense to also move these lines up to the `ast.literal_eval`? And also move the `try` into the `if`? That's what it's there for, right?

sebageek replied:
The method is only copied from `driver.py` over to this file, I didn't change much in it. If we were to improve it what do you think about tossing out `ast.literal_eval`, use `json.loads` and catch that value error directly there? I'd probably do an extra commit for this though.

joker-at-work replied:
I'd prefer `json.loads` and yes, catch it where it's raised.

joker-at-work commented on networking_aci/plugins/ml2/drivers/mech_aci/config.py:
```diff
@@ -193,6 +199,12 @@ def _parse_fixed_bindings(self):
 
+        for scope in self._address_scopes:
+            if scope.get('contracts'):
+                contracts = ast.literal_eval(scope.get('contracts'))
+                scope['consumed_contracts'].extend(contracts['consumed'])
+                scope['provided_contracts'].extend(contracts['provided'])
```
How bad would it be, if contracts are duplicated in the `consumed_contracts` or `provided_contracts`?

sebageek replied:
We might push them two times to the APIC, but it wouldn't cause a problem

sebageek committed:
Unroll _get_networks()

Less indentation, makes it a bit more readable for me.

sebageek committed:
Put getting host from port into utility function

get_host_from_profile() grabs the host specified in the
local_link_information, if present. If not present it uses the host
provided by the port.

sebageek committed:
Fix parameter indentation

sebageek committed:
Direct-on-aci portbindings with mode-switching API

Direct-mode is a new config option for a hostgroup. A hostgroup in
direct mode is expected to be directly bound by the driver. For these
hosts we still do an initial hierarchical portbinding but bind said port
in the second bind_port() call.

Direct-mode hostgroups can have two modes, which change how a port in
this port is being bound:

MODE_INFRA: A port in this group will be bound like a normal VM port,
except that on the ACI side the encapsulation mode will be access /
untagged instead of trunk / tagged. This mode can be used to install a
hypervisor host via ironic. A MODE_INFRA hostgroup will share a physical
domain and vlan-pool with a parent hostgroup. This parent hostgroup is
is used for regular VMs, is not a direct-mode hostgroup and will result
in normal tagged bindings (as always). A portbinding of a port in
MODE_INFRA will take precedence, i.e. its bindings will be deployed as
untagged in an EPG if deployed on a network where a port in the parents
hostgroup is already deployed.

MODE_BAREMETAL: A port in this segment will be bound on its own segment,
have its own physical domain and vlan pool. A custom VPC policy group
will be created and all port selectors referenced in this hostgorup will
be assigned to this policy group. Ports will be bound untagged. Bindings
of a hostgroup in MODE_BAREMETAL will be removed from its parent group.

An API for the ACI driver is now present. It allows to get insight in
the current config (more of an API test endpoint, but might be useful in
the future).
This API also contains means to list and set the hostgroup mode of all
direct-mode hostgroups. On switching the ACI agent will be notified to
perform changes on the ACI, switching interface profiles etc..

To better support these new modes the rpc api support call
_get_network() was refactored. It now gets all hosts on a network and
adds hostgroup-based config accordingly to its response. This should
result in a faster call and less code to maintain.

sebageek committed:
Add trunk support for MODE_BAREMETAL

The user can now create trunks and subports for baremetal hosts
(meaning hosts in an hostgroup that is in MODE_BAREMETAL). The user can
attach OpenStack networks with vlan ids of their choosing as trunk ports
to a server. The separate physical domain in ACI already used by
MODE_BAREMETAL allows us to use a completely separate vlan pool.

sebageek commented:
The Direct-on-aci branch is still missing its migrations for the aci_hostgroup_mode table. I will add them for review once we are done with the rest of this branch.

sebageek committed:
Assume direct_mode false for hostgroups in agent

Fixed bindings don't provide a direct-mode (it is not applicable for
them). Instead of adding a direct_mode entry in their config dict we
just default to False in the agent for direct_mode. This has also the
nice side effect that on the transition from non-direct_mode aware
neutron-server instances to direct_mode-aware we don't get any
exceptions regarding missing keys.

notandy commented on networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py:
```diff
@@ -207,39 +201,76 @@ def _release_vxlan_segment(self, network, host_config, level, segment):
+        network_id = network['id']
+
+        session = db_api.get_writer_session()
+        segment = session.query(ml2_models.NetworkSegment).filter_by(segmentation_id=segmentation_id,
```
I know it's implemented similar as before, but wouldn't it make sense to use with_for_update() to lock the segment table? Thus you can ensure that no segments are parallel created by different instances of neutron.

notandy replied:
Ah probably this will be catched by the table constraints

sebageek replied:
For baremetal segments we don't have an extra VLAN pool so the constraints will not protect us - I think. What we would need would be constraints on the networksegments table and that would be a bigger endeavour. The other solution would be to create an extra VLAN pool for all possible ~4095 minus blacklisted vlan ids per node, which would be quite a lot.

On the other hand... we will only have one node per hostgroup and with that only one port allocated to it per network and segment id (with portgroups being used for this, of course). So we should not end up in a race-condition-scenario here.

notandy commented on networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py:
```diff
@@ -305,3 +339,20 @@ def _process_host_or_hostgroup(self, session, host, host_config, allocations, le
+        with session.begin(subtransactions=True):
+            # fetch all mode-hostgroups from db
+            db_groups = []
+            for db_entry in (session.query(HostgroupModeModel).with_lockmode('update')):
```
with_lockmode seems to be depracted since sqlalchemy 0.9: https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.with_lockmode

notandy reviewed (approved):
👍

sebageek committed:
Replace deprecated with_lockmode()

SQLAlchemy has deprecated with_lockmode(), we can replace it with
with_for_update(). See docs:

https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.with_lockmode

sebageek committed:
Add a bit more logging for segment management

joker-at-work commented on networking_aci/extensions/acioperations.py:
```diff
@@ -0,0 +1,219 @@
+            'fixed_bindings': ACI_CONFIG.fixed_bindings,
+            'address_scopes': ACI_CONFIG.address_scopes,
+        }
+        # return {"status": "success", "index": True, "arguments": kwargs}
```
Seems like this can go now that this functions is implemented :D

joker-at-work commented on networking_aci/extensions/acioperations.py:
```diff
@@ -0,0 +1,219 @@
+
+    def index(self, request, **kwargs):
+        """Show all hostgroup, fixed bindings and address scopes"""
+        check_cloud_admin(request)
```
Would look better as a decorator imho.

joker-at-work commented on networking_aci/extensions/acioperations.py:
```diff
@@ -0,0 +1,219 @@
+        if curr_mode == new_mode:
+            raise web_exc.HTTPBadRequest("{} is already in {} mode".format(hostgroup_name, new_mode))
+
+        # check if there are still  on this hostgroup
```
I think there's a word missing in this comment (is it "segment" or "portbinding"?) - or a space too much.

joker-at-work commented on networking_aci/extensions/acioperations.py:
```diff
@@ -0,0 +1,219 @@
+            raise web_exc.HTTPBadRequest("Hostgroup {} has no config associated with it".format(hostgroup_name))
+
+        for segment_id in self.db.get_segment_ids_by_physnet(ctx, hg_config['physical_network']):
+            hosts = self.db.get_hosts_on_segment(ctx, segment_id)
+            for host in hg_config['hosts']:
+                if host in hosts:
+                    LOG.error("Denied API request to switch hostgroup %s to %s mode, host %s still bound in segment %s",
+                              hostgroup_name, new_mode, host, segment_id)
```
If you use set operations, you could return all the hosts being still problematic in a single reply, making the RTT for freeing up the hostgroup shorter. But just a suggestion and maybe not worth the effort/follow-up work.

sebageek replied:
In nearly all cases we only will have one host in the hostgroup for direct_mode hostgroups, so I don't it's worth adding a "host(s)" to the commit message and making it therefore less beautiful. ;)

joker-at-work commented on networking_aci/extensions/acioperations.py:
```diff
@@ -0,0 +1,219 @@
+                self.rpc_notifier.clean_baremetal_objects(hg_config)
+                baremetal_bindings_cleaned = True
+            except Exception:
+                LOG.exception("Could not clean baremetal bindings from ACI, cleaning will have to be done manually")
```
The client will not see that something went wrong. Can we provide the info somehow in an additional field? Might help in automating this via ansible later on.
Or does this need domain-knowledge in the automation, that we need to check `baremetal_bindings_cleaned` when setting away from `MODE_BAREMETAL`? Fine, too.

sebageek replied:
Jup, this is hard domain knowledge. The flag `baremetal_bindings_cleaned` is only interesting in the switchover case from baremetal to infrastructure mode (else there are no bindings to clean). If the API returns false then there was a cleanup error.

joker-at-work commented on networking_aci/extensions/acioperations.py:
```diff
@@ -0,0 +1,219 @@
+                "baremetal_bindings_cleaned": baremetal_bindings_cleaned,
+                "aci_objects_update_succeeded": aci_objects_update_succeeded,
+            }
+        else:
+            LOG.error("Failed to set hostgroup %s to mode %s - no DB entry?", hostgroup_name, new_mode)
+            raise web_exc.HTTPConflict("Could not set hostmode of hostgroup %s - DB entry could not be found",
+                                       hostgroup_name)
```
Going here, we might have cleaned up baremetal objects already, while still staying in `MODE_BAREMETAL` - e.g. if the DB went away or something. Can we catch that case somehow?

sebageek replied:
We could try to do another apicall to reset everything, but for switch from baremetal to infra this would not be problem. For infra to baremetal the host would not be usable until it's synced again. I'm not sure if this is a problem we'll ever run into, but manual intervention would somehow be required.

So... not sure about my options here. I could make the error mesage more explicit, maybe also only for the infra --> baremetal case or I could try to do another API call to the APIC to reset everything. Seems like a lot of work for this though.

joker-at-work replied:
I've seen code down the line that does an "ensure_baremetal_objects" or something. If that's called every time we try to make a portbinding, then it should still work if we just cleaned and then ended up not switching the mode, right?

sebageek replied:
Everything is synced again on the first deployment to a hostgroup, yes. The baremetal cleanup only removes unused API objects from the APIC.

joker-at-work commented on networking_aci/extensions/acioperations.py:
```diff
@@ -0,0 +1,219 @@
+                LOG.exception("Could not update bindings on ACI, rpc call failed")
+
+            return {
+                "success": True,
```
We always return non-JSON error messages when something broke along the line, so I would assume a status-code of 200 would be enough for success and this additional key is basically useless, because it can only be there, if we succeeded.

sebageek replied:
I added this for the case that someone using the API would like to check for success in a `if "success" in req.json()` manner. Do you think this is a bad idea?

joker-at-work replied:
Not necessarily. I'd prefer if they'd check the HTTP status-code though.
Doesn't that type of pattern break for every error we return or does Neutron convert those errors into JSON, too?

sebageek replied:
Neutron converts (most) errors to JSON as well

joker-at-work commented on networking_aci/extensions/acioperations.py:
```diff
@@ -0,0 +1,219 @@
+
+            return {
+                "success": True,
+                "baremetal_bindings_cleaned": baremetal_bindings_cleaned,
+                "aci_objects_update_succeeded": aci_objects_update_succeeded,
```
Just for my understand: How important are those two? If one of them is `False`, does that mean "successful, but you cannot work with that until checked manually"? If yes, how would the next automation request identify that this is the case?
"We need some follow-ups to make this work with automation. This is for manual use for now." is a valid answer here, too.

sebageek replied:
In case this breaks we probably need manual intervention, yes. It should be usable by automation, but if it breaks we need to check what's what (at least for now).

joker-at-work replied:
I fear that people will just retry and neither Neutron nor ansible having state between jobs will think everything is fine ... :/

Maybe instead of bailing out with an error, if someone sets the same mode the hostgroup currently has, we could try to check if cleanup is still necessary? Basically making this command idempotent. It always returns success, if the wanted state is achieved.

sebageek replied:
To clarify: The configuration is done again when the first infrastructure or baremetal machine is deployed in the hostgroup that is switched. It will then be done by the syncloop in periodic intervals als well. The only way to break this is by switching from baremetal to infra, then it needs to break on the ACI and then you deploy the machine by hand (ie without using ironic). So I think we are fine here for now. I think we'll approach this if this use-case arises.

joker-at-work commented on networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py:
```diff
@@ -207,39 +201,78 @@ def _release_vxlan_segment(self, network, host_config, level, segment):
+                )
+                session.add(segment)
+        else:
+            # check segment has corrent id
```
"corrent" -> "correct"?

joker-at-work commented on networking_aci/plugins/ml2/drivers/mech_aci/common.py:
```diff
@@ -71,22 +75,121 @@ def get_address_scope_name(self, context, subnet_pool_id):
+        query = query.join(ml2_models.PortBindingLevel,
+                           ml2_models.PortBindingLevel.segment_id == segment_models.NetworkSegment.id)
+        query = query.group_by(segment_models.NetworkSegment.network_id)
+        query = query.filter(segment_models.NetworkSegment.physical_network == 'bb92')
```
Why is `bb92` hardcoded here?

sebageek replied:
Wow, that is a somewhat big oversight. This method seems to be used nowhere anyway, so I'm going to remove it.

sebageek committed:
Replace literal_eval with json for binding_profile

JSON objects look similar enough to python pickle objects so that they
can be loaded with ast.literal_eval, though there is no reason in our
case to do so. OpenStack itself also handles the binding_profile as a
JSON objects, so we can do this here as well.

sebageek committed:
Add gitignore

notandy commented:
🎊
